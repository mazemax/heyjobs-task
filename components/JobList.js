import React from 'react';
import Link from 'next/link';
import PropTypes from 'prop-types';
import { useDispatch } from 'react-redux';

import { setJob } from '../store/home/action';
import {
    Avatar, LinkWrapper, List, ListItem, ListItemIcon, ListItemText, ListItemSecondaryAction,
    IconButton, Divider, BusinessCenterSharpIcon, ArrowForwardIosIcon
} from '../styles/JobList.style';

/**
 * Functional component for Job listing
 * @param data
 * @returns {*}
 * @constructor
 */
function JobList ({ data }) {
    const dispatch = useDispatch();

    return (
        <List component="nav" aria-label="main">

            {
                data.map(job => (
                    <React.Fragment key={job.id}>
                        <Link href="/detail">
                            <LinkWrapper onClick={() => dispatch(setJob(job))}>
                                <ListItem button>
                                    <ListItemIcon>
                                        <Avatar variant="rounded">
                                            <BusinessCenterSharpIcon color={'secondary'} />
                                        </Avatar>
                                    </ListItemIcon>
                                    <ListItemText primary={job.title} />
                                    <ListItemSecondaryAction>
                                        <IconButton edge="end" aria-label="delete">
                                            <ArrowForwardIosIcon />
                                        </IconButton>
                                    </ListItemSecondaryAction>
                                </ListItem>
                            </LinkWrapper>
                        </Link>
                        <Divider />
                    </React.Fragment>
                ))
            }

        </List>
    );
}

/**
 * Typechecking for JobList
 * @type {{data: *}}
 */
JobList.propTypes = {
    /**
     * Job data retrieved from API
     */
    data: PropTypes.array.isRequired
};

export default JobList;
