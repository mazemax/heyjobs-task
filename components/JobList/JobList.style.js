import styled from 'styled-components';
import MaterialAvatar from '@material-ui/core/Avatar';
import MaterialList from '@material-ui/core/List';
import MaterialListItem from '@material-ui/core/ListItem';
import MaterialListItemIcon from '@material-ui/core/ListItemIcon';
import MaterialListItemText from '@material-ui/core/ListItemText';
import MaterialListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import MaterialDivider from '@material-ui/core/Divider';
import MaterialBusinessCenterSharpIcon from '@material-ui/icons/BusinessCenterSharp';
import MaterialIconButton from '@material-ui/core/IconButton';
import MaterialArrowForwardIosIcon from '@material-ui/icons/ArrowForwardIos';

// Note: Over-ridding Material UI Elements using Styled Components

export const LinkWrapper = styled.a`
  text-decoration: none;
  color: initial;
`;

export const Avatar = styled(MaterialAvatar)`
  && {
    background-color: #d8e0ed;
    }
`;

export const List = styled(MaterialList)`
  && {
  }
`;

export const ListItem = styled(MaterialListItem)`
  && {
  }
`;

export const ListItemIcon = styled(MaterialListItemIcon)`
  && {
  }
`;

export const ListItemText = styled(MaterialListItemText)`
  && {
  }
`;

export const ListItemSecondaryAction = styled(MaterialListItemSecondaryAction)`
  && {
  }
`;

export const Divider = styled(MaterialDivider)`
  && {
  }
`;

export const BusinessCenterSharpIcon = styled(MaterialBusinessCenterSharpIcon)`
  && {
  }
`;

export const IconButton = styled(MaterialIconButton)`
  && {
  }
`;

export const ArrowForwardIosIcon = styled(MaterialArrowForwardIosIcon)`
  && {
  }
`;
