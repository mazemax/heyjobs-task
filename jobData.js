/**
 * Hard-coded data for job listing
 * @type {*[]}
 */
export const jobs = [
    {
        id: '1',
        title: 'Frontend Developer',
        description: 'Comfortable with modern JS stack, experience with React.',
        employment_type: 'full_time'
    },
    {
        id: '2',
        title: 'Backend Developer',
        description: 'Comfortable with modern Ruby stack, experience with Ruby on Rails.',
        employment_type: 'part_time'
    }
];
