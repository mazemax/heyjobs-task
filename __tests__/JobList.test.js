/* eslint-env jest */

import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';

import JobList from '../components/JobList';
import { jobs } from '../jobData';

jest.mock('next/link', () => {
    return ({children}) => {
        return children;
    }
});

const initialState = {
    home: {
        id: '0',
        title: '',
        description: '',
        employmentType: ''
    }
};
const mockStore = configureStore();
const store = mockStore(initialState);

describe('Testing JobList component', () => {
    test('Should trigger SET_JOB action in redux store correctly', async () => {
        const props = {
            data: jobs
        };
        const component = mount(
            <Provider store={store}>
                <JobList {...props} />
            </Provider>
        );

        await component.find('a').first().simulate('click');
        expect(store.getActions()).toMatchSnapshot();
    });
});
