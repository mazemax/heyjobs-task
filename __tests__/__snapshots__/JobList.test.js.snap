// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing JobList component Should trigger SET_JOB action in redux store correctly 1`] = `
Array [
  Object {
    "payload": Object {
      "description": "Comfortable with modern JS stack, experience with React.",
      "employment_type": "full_time",
      "id": "1",
      "title": "Frontend Developer",
    },
    "type": "SET_JOB",
  },
]
`;
